source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -48,66 +50,104 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
-  if (request_state_ == FilterState::HEADERS) {
-    if (response->has_request_headers()) {
-      ENVOY_LOG(debug, "applying request_headers response");
-      message_valid = true;
-      const auto& headers_response = response->request_headers();
-      if (headers_response.has_response()) {
-        const auto& common_response = headers_response.response();
-        if (common_response.has_header_mutation()) {
-          MutationUtils::applyHeaderMutations(common_response.header_mutation(), *request_headers_);
-        }
-      }
-    } else if (response->has_immediate_response()) {
-      ENVOY_LOG(debug, "Returning immediate response from processor");
-      sendImmediateResponse(response->immediate_response());
-      message_valid = true;
-    }
-    request_state_ = FilterState::IDLE;
-    decoder_callbacks_->continueDecoding();
+  if (response->has_request_headers()) {
+    message_handled = handleRequestHeadersResponse(response->request_headers());
+  } else if (response->has_response_headers()) {
+    message_handled = handleResponseHeadersResponse(response->response_headers());
+  } else if (response->has_immediate_response()) {
+    handleImmediateResponse(response->immediate_response());
+    message_handled = true;
   }
 
-  if (message_valid) {
+  if (message_handled) {
     stats_.stream_msgs_received_.inc();
   } else {
     stats_.spurious_msgs_received_.inc();
     // Ignore messages received out of order. However, close the stream to
     // protect ourselves since the server is not following the protocol.
     ENVOY_LOG(warn, "Spurious response message received on gRPC stream");
+    cleanupState();
     closeStream();
   }
 }
 
+bool Filter::handleRequestHeadersResponse(const HeadersResponse& response) {
+  if (request_state_ == FilterState::HEADERS) {
+    ENVOY_LOG(debug, "applying request_headers response");
+    MutationUtils::applyCommonHeaderResponse(response, *request_headers_);
+    request_state_ = FilterState::IDLE;
+    decoder_callbacks_->continueDecoding();
+    return true;
+  }
+  return false;
+}
+
+bool Filter::handleResponseHeadersResponse(const HeadersResponse& response) {
+  if (response_state_ == FilterState::HEADERS) {
+    ENVOY_LOG(debug, "applying response_headers response");
+    MutationUtils::applyCommonHeaderResponse(response, *response_headers_);
+    response_state_ = FilterState::IDLE;
+    encoder_callbacks_->continueEncoding();
+    return true;
+  }
+  return false;
+}
+
+void Filter::handleImmediateResponse(const ImmediateResponse& response) {
+  // We don't want to process any more stream messages after this.
+  // Close the stream before sending because "sendLocalResponse" triggers
+  // additional calls to this filter.
+  request_state_ = FilterState::IDLE;
+  response_state_ = FilterState::IDLE;
+  closeStream();
+
+  if (encoding_started_) {
+    ENVOY_LOG(debug, "Returning immediate response from processor on encoding path");
```

Comment:
Once encoding has actually started, I don't think we can do any local reply any more.

Reply:
The comments in filter.h say this:

"If a response has already started (e.g. if the router calls sendSendLocalReply after encoding headers) this will either ship the reply directly to the downstream codec, or reset the stream."

This definitely works after encodeHeaders has been called and before we call continueEncoding. After that we don't have many options, so we either call this or we somehow replace the response body ourselves. When we get to implementing body handling I planned to write integration tests to make sure we understand what Envoy will actually do.